### PR TITLE
Keep WithNonVoidReturnType from matching void

### PR DESCRIPTION
**Pull request by blairconrad**
An evolution of #967, but targeting our 2.x stream.

I've incorporated @thomaslevesque's suggestions from #967, specifically adding a commit to ensure that `WithNonVoidReturnType` rules describe themselves properly, and making `OnIsApplicableTo` more readable (I deviated a little from the suggested implementation, as I prefer early returns.)

### CODE REVIEW

**Comment by blairconrad** [comment]
There we go, @thomaslevesque. I'm happy to remove the early return in `OnIsApplicableTo`, if you prefer. Or to squash anything you want, or whatever.

File: src/FakeItEasy/Configuration/AnyCallCallRule.cs [hunk 1]
```diff
@@ -37,11 +37,22 @@ public override void UsePredicateToValidateArguments(Func<ArgumentCollection, bo
...
+
+            if (this.ApplicableToAllNonVoidReturnTypes)
+            {
+                return fakeObjectCall.Method.ReturnType != typeof(void);
+            }
+
+            if (this.ApplicableToMembersWithReturnType != null)
+            {
+                return this.ApplicableToMembersWithReturnType == fakeObjectCall.Method.ReturnType;
+            }
```
Comment (thomaslevesque) [inline src/FakeItEasy/Configuration/AnyCallCallRule.cs]: I'm wondering... can we have a situation where both `ApplicableToAllNonVoidReturnTypes` and `ApplicableToMembersWithReturnType != null` are true? I don't think so, but assuming it's possible, shouldn't we check `ApplicableToMembersWithReturnType` first, as it's more specific?

File: src/FakeItEasy/Configuration/AnyCallCallRule.cs [hunk 2]
```diff
@@ -21,6 +21,11 @@ public override string DescriptionOfValidCall
         {
             get
             {
+                if (this.ApplicableToAllNonVoidReturnTypes)
+                {
+                    return "Any call with non-void return type to the fake object.";
+                }
```
Comment (thomaslevesque) [inline src/FakeItEasy/Configuration/AnyCallCallRule.cs]: Same concern here as in `OnIsApplicableTo`

File: src/FakeItEasy/Configuration/AnyCallCallRule.cs (hunk 1 above)
Comment (blairconrad) [inline src/FakeItEasy/Configuration/AnyCallCallRule.cs]: I don't think that we can be in such a situation, but I like your idea to change the order of the checks. Fixup commits incoming.

**Comment by thomaslevesque** [comment]
LGTM! I'll merge after you squash the fixup commits

**Comment by blairconrad** [comment]
Squashed and pushed. Thanks, @thomaslevesque, vastly improving the content.

**Comment by thomaslevesque** [comment]
Thanks @blairconrad !

**Pull request merged**